src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,321 @@
+    static int rIndex = 0;
+
+    assert (rIndex < 128);
+
+    if (iterations >= numUpdates / 128){
+        doneReading = true;
+
+        // exitSimLoop(name() + " has encountered the exit state and will "
+                // "terminate the simulation.\n");
+        return;
+    }
+
+    randomSeeds[rIndex] = (randomSeeds[rIndex] << 1) ^
+                    ((int64_t) randomSeeds[rIndex] < 0 ? POLY : 0);
+    uint64_t value = randomSeeds[rIndex];
+    uint64_t index = randomSeeds[rIndex] & (tableSize - 1);
+    Addr addr = indexToAddr(index);
+    updateTable[addr] = value;
+    PacketPtr pkt = getReadPacket(addr, sizeof(uint64_t));
+    DPRINTF(GUPSGen, "generateNextBatch: Pushed pkt, pkt->addr_range: %s,"
+                    "(pkt != nullptr): %d\n",
+                    pkt->getAddrRange().to_string(), pkt != nullptr);
+    requestPool.push(pkt);
+
+    rIndex++;
+    if (rIndex >= 128){
+        rIndex = 0;
+        iterations++;
+    }
+
+    schedule(nextGenEvent, curTick() + 250);
+
+    if (!nextSendEvent.scheduled() && !port.blocked()
+                            && !requestPool.empty()) {
+        schedule(nextSendEvent, curTick() + 100);
+    }
+}
+
+
+void
+GUPSGen::sendNextBatch()
+{
+    if (port.blocked()){
+        return;
+    }
+
+    if (!requestPool.empty())
+    {
+        PacketPtr pkt = requestPool.front();
+
+        exitTimes[pkt] = curTick();
+        DPRINTF(GUPSGen, "sendNextBatch: Sent pkt, pkt->addr_range: %s,"
+                        "(pkt != nullptr): %d\n",
+                        pkt->getAddrRange().to_string(), pkt != nullptr);
+        port.sendTimingPacket(pkt);
+        onTheFlyRequests++;
+        requestPool.pop();
+    }
+
+    if (!nextSendEvent.scheduled() && !port.blocked()
+                                && !requestPool.empty()) {
+        schedule(nextSendEvent, curTick() + 250);
+    }
+
+    // if (requestPool.empty())
+    // {
+    //     generateNextBatch();
+    // }
+
+    return;
+}
+
+PacketPtr
+GUPSGen::getReadPacket(Addr addr, unsigned int size)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size,
+                                            0, requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    return pkt;
+}
+
+PacketPtr
+GUPSGen::getWritePacket(Addr addr, unsigned int size, uint8_t *data)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size, 0,
+                                               requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::WriteReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
```

Comment:
Would it make sense to actually write some "real" dummy data and check that data on reads? This could check the functionality of memory.

Do the other traffic gens do this?

Reply:
I almost did the same thing as other generators. In the startup function I actually do write some data into the memory, I am going to use the same function to check for that too.

Reply:
Thanks for the tip